cpp/dolfinx/fem/discreteoperators.h

```diff
@@ -7,56 +7,54 @@
 #pragma once
 
 #include "DofMap.h"
+#include "FiniteElement.h"
 #include "FunctionSpace.h"
 #include <array>
 #include <dolfinx/common/IndexMap.h>
+#include <dolfinx/common/math.h>
 #include <dolfinx/la/SparsityPattern.h>
 #include <dolfinx/mesh/Mesh.h>
 #include <memory>
 #include <vector>
+#include <xtensor/xadapt.hpp>
+#include <xtensor/xbuilder.hpp>
+#include <xtensor/xtensor.hpp>
 #include <xtl/xspan.hpp>
 
 namespace dolfinx::fem
 {
 
-/// @todo Improve documentation
-/// This function class computes the sparsity pattern for discrete gradient
-/// operators (matrices) that map derivatives of finite element functions into
-/// other finite element spaces.
+/// @brief Create a sparsity pattern for a discrete gradient
+/// operator.
 ///
-/// @warning This function is highly experimental and likely to change
-/// or be replaced or be removed
+/// @warning This function relies on the user supplying appropriate
+/// input and output spaces. See parameter descriptions.
 ///
-/// Build the sparsity for the discrete gradient operator A that takes a
-/// \f$w \in H^1\f$ (P1, nodal Lagrange) to \f$v \in H(curl)\f$
-/// (lowest order Nedelec), i.e. v = Aw. V0 is the H(curl) space,
-/// and V1 is the P1 Lagrange space.
-///
-/// @param[in] V0 A H(curl) space
-/// @param[in] V1 A P1 Lagrange space
-/// @return The sparsity pattern
+/// @param[in] V0 A q-th order Nedelec (first kind) space
+/// @param[in] V1 A p-th order Lagrange space
+/// @param[in] V0 A degree q Nedelec (first kind) space
+/// @param[in] V1 A degere p Lagrange space
 la::SparsityPattern
```

Comment:
This seems like just a re-implementation of `fem::sparsitybuild::cells`, and it doesn't see to exploit any special non-zero structure for the discrete gradient problem.